src/gun_http2.erl

```diff
@@ -1074,63 +1087,76 @@ maybe_send_data(State=#http2_state{http2_machine=HTTP2Machine0}, StreamID, IsFin
 	end,
 	case cow_http2_machine:send_or_queue_data(StreamID, HTTP2Machine0, IsFin, Data) of
 		{ok, HTTP2Machine} ->
-			{State#http2_state{http2_machine=HTTP2Machine}, EvHandlerState};
+			{{state, State#http2_state{http2_machine=HTTP2Machine}}, EvHandlerState};
 		{send, SendData, HTTP2Machine} ->
 			send_data(State#http2_state{http2_machine=HTTP2Machine}, SendData,
 				EvHandler, EvHandlerState)
 	end.
 
 send_data(State, [], _, EvHandlerState) ->
-	{State, EvHandlerState};
+	{{state, State}, EvHandlerState};
 send_data(State0, [{StreamID, IsFin, SendData}|Tail], EvHandler, EvHandlerState0) ->
-	{State, EvHandlerState} = send_data(State0, StreamID, IsFin, SendData, EvHandler, EvHandlerState0),
-	send_data(State, Tail, EvHandler, EvHandlerState).
+	case send_data(State0, StreamID, IsFin, SendData, EvHandler, EvHandlerState0) of
+		{{state, State}, EvHandlerState} ->
+			send_data(State, Tail, EvHandler, EvHandlerState);
+		ErrorResult={{error, _}, _EvHandlerState} ->
+			ErrorResult
+	end.
 
 send_data(State0, StreamID, IsFin, [Data], EvHandler, EvHandlerState0) ->
-	State = send_data_frame(State0, StreamID, IsFin, Data),
-	EvHandlerState = case IsFin of
-		nofin ->
-			EvHandlerState0;
-		fin ->
-			#stream{ref=StreamRef, reply_to=ReplyTo} = get_stream_by_id(State, StreamID),
-			RequestEndEvent = #{
-				stream_ref => stream_ref(State, StreamRef),
-				reply_to => ReplyTo
-			},
-			EvHandler:request_end(RequestEndEvent, EvHandlerState0)
-	end,
-	{maybe_delete_stream(State, StreamID, local, IsFin), EvHandlerState};
+	case send_data_frame(State0, StreamID, IsFin, Data) of
+		{state, State} ->
+			EvHandlerState = case IsFin of
+				nofin ->
+					EvHandlerState0;
+				fin ->
+					#stream{ref=StreamRef, reply_to=ReplyTo} = get_stream_by_id(State, StreamID),
+					RequestEndEvent = #{
+						stream_ref => stream_ref(State, StreamRef),
+						reply_to => ReplyTo
+					},
+					EvHandler:request_end(RequestEndEvent, EvHandlerState0)
+			end,
+			{{state, maybe_delete_stream(State, StreamID, local, IsFin)}, EvHandlerState};
+		Error={error, _Reason} ->
+			{Error, EvHandlerState0}
```

Comment:
As a counter-example to the previous comment, here if we fail to send the final data frame it makes sense to not update the state. So it's OK to keep the maybe_delete_stream for only successful sends as you did.

Reply:
No changes needed here.